### ISSUE TITLE
Redirecting on button click

### ISSUE THREAD

**Issue by trentrand** [tomasbjerre/pull-request-notifier-for-bitbucket#307]
I'd like to use this plugin to make a request, then use part of the response in order to formulate a URL that the user is then redirected to.

It does not seem like this plugin supports redirection. Is there any way to achieve this?

**Comment by tomasbjerre** [comment]
It can be implemented.

**Comment by trentrand** [comment]
I'd be willing to contribute this feature, if you don't have time in the near future.

Would appreciate some direction on where you'd like this field to live.

I'm considering adding this as an additional field to the bottom of the notifications panel:
<img width="644" alt="screen shot 2018-11-10 at 2 01 10 pm" src="https://user-images.githubusercontent.com/6083067/48304985-20012f80-e4f1-11e8-9306-69b869f7d5f4.png">

**Comment by tomasbjerre** [comment]
Contributions are very welcome! As long as they are not very complex and in need of much maintenance in the future.

Perhaps having a redirect url in the button config. Let it use variables in the url. Have config in the notification that lets user extract values from rrsponse into variables. Perhaps using jsonpath or xpath to extract values...

**Comment by trentrand** [comment]
Would you be able to publish it Atlassian Server marketplace shortly after the feature is released?

**Comment by tomasbjerre** [comment]
Yes.

**Comment by trentrand** [comment]
Does it make more sense to include this field in the Notification URL field? 

Otherwise, it is not possible to create a dynamic redirect using the `${INJECTION_URL_VALUE}`  variable without adding some arbitrary value to the Notification URL field.

If the field was moved there, the condition for creating / updating a notification could enforce the inclusion of either the Notification `Url` or `Redirect Url`.

Alternative solutions may instead be:
- Remove requirement for specifying a Notification `Url`
- Remove visibility requirement from buttons:
  - > Only buttons that will trigger at least one configured notification is shown

**Comment by tomasbjerre** [comment]
Having redirect url as part of notification would be confusing for anyone not using buttons. I think it belongs with the button config as that is the only use case for it.

Perhaps redirect when only one notification matches so that you can use its injection url value.

Ofcourse you have the option to fork this repo and do this in you fork without merging back. If your organization lets you install custom plugins. This option may be preffered if it cannot be implemented in a very user friendly way. I dont have much time to spend on maintenance.

**Comment by trentrand** [comment]
I've created a straight-forward implementation that works well; I'm just struggling to come up with the best location for rendering variables into the Redirect Url.

Right now I'm thinking of doing-so inside of the PrnfbButton.java `Redirect Url` getter using the following snippet:

import se.bjurr.prnfb.service.PrnfbRenderer.ENCODE_FOR;

```
final PrnfbRendererWrapper renderer =
          buttonsService.getRenderer(repositoryId, pullRequestId, dto.getUuid());
final String redirectUrl = buttonFormElementDto.getRedirectUrl();
final String redirectUrlRendered = renderer.render(redirectUrl, ENCODE_FOR.URL);
```

I'm looking into how I can get those renderer initialization values into this class; or how I can inject an initialized renderer.

Any guidance?

**Comment by tomasbjerre** [comment]
The `PrnfbButton` should just be a data class for storing settings. No such logic allowed.

Do rendering with `public String render(String inputString, ENCODE_FOR encodeFor)` in `PrnfbRendererWrapper`.

If did not add any extra variables, only using `INJECTION_URL_VALUE`, then this is probably all you need to do. Just use the configured `redirectUrl` as `inputString`.

**Comment by trentrand** [comment]
So if the variable rendering shouldn't be done inside of `PrnfbButton`, where should it be done?
`ButtonDTO`? `ButtonTransformer`?

**Comment by tomasbjerre** [comment]
No. DTO is just for data transfer. Transformer for transforming to another type without changing data.

Read my whole previous comment.

**Comment by tomasbjerre** [comment]
Released in 3.24 now.
